src/voter/mod.rs

```diff
@@ -1168,4 +1241,105 @@ mod tests {
 			}).map(move |_| signal.fire())
 		})).unwrap();
 	}
+
+	#[test]
+	fn pick_up_from_prior_without_grandparent_state() {
+		let local_id = Id(5);
+		let voters = std::iter::once((local_id, 100)).collect();
+
+		let (network, routing_task) = testing::environment::make_network();
+		let (signal, exit) = ::exit_future::signal();
+
+		let global_comms = network.make_global_comms();
+		let env = Arc::new(Environment::new(network, local_id));
+		current_thread::block_on_all(::futures::future::lazy(move || {
+			// initialize chain
+			let last_finalized = env.with_chain(|chain| {
+				chain.push_blocks(GENESIS_HASH, &["A", "B", "C", "D", "E"]);
+				chain.last_finalized()
+			});
+
+			// run voter in background. scheduling it to shut down at the end.
+			let finalized = env.finalized_stream();
+			let voter = Voter::new(
+				env.clone(),
+				voters,
+				global_comms,
+				10,
+				Vec::new(),
+				last_finalized,
+				last_finalized,
+			);
+			::tokio::spawn(exit.clone()
+				.until(voter.map_err(|_| panic!("Error voting"))).map(|_| ()));
+
+			::tokio::spawn(exit.until(routing_task).map(|_| ()));
+
+			// wait for the best block to finalize.
+			finalized
+				.take_while(|&(_, n, _)| Ok(n < 6))
+				.for_each(|_| Ok(()))
+				.map(|_| signal.fire())
+		})).unwrap();
+	}
+
+	#[test]
+	fn pick_up_from_prior_with_grandparent_state() {
+		let local_id = Id(5);
+		let voters = std::iter::once((local_id, 100)).collect();
+
+		let (network, routing_task) = testing::environment::make_network();
+		let (signal, exit) = ::exit_future::signal();
+
+		let global_comms = network.make_global_comms();
+		let env = Arc::new(Environment::new(network, local_id));
+		current_thread::block_on_all(::futures::future::lazy(move || {
+			// initialize chain
+			let last_finalized = env.with_chain(|chain| {
+				chain.push_blocks(GENESIS_HASH, &["A", "B", "C", "D", "E"]);
+				chain.last_finalized()
+			});
+
+			env.note_completed_round(9, RoundState::genesis(("C", 4)));
+
+			// run voter in background. scheduling it to shut down at the end.
+			let finalized = env.finalized_stream();
+			let voter = Voter::new(
+				env.clone(),
+				voters,
+				global_comms,
+				10,
+				vec![
+					SignedMessage {
+						message: crate::Message::Prevote(crate::Prevote {
+							target_hash: "D",
+							target_number: 5,
+						}),
+						signature: Signature(5),
+						id: local_id,
+					},
+					SignedMessage {
+						message: crate::Message::Precommit(crate::Precommit {
+							target_hash: "D",
+							target_number: 5,
+						}),
+						signature: Signature(5),
+						id: local_id,
+					},
+				],
+				last_finalized,
```

Comment:
This will be `(Genesis, 0)` if I'm not mistaken, but the grandparent round finalized `(C, 4)`. We should make a test where we actually need to have the background running in order to progress to the latest round (i.e. we need votes on the last round to push the estimate backwards).